File: peerrpc-go/signal/ws.go
// Package signal WebSocket backend: a signaling client that speaks
// the signaling wire format (service / AnnounceRequest) to a remote
// signal-server over a raw WebSocket.
//
// The frame format matches signalserver/ws.go exactly: each
// WebSocket BinaryMessage is a 4-byte big-endian length prefix
// followed by a marshaled peerrpc.signaling.SignalMessage protobuf.
// The first frame the client sends MUST be an AnnounceRequest.

package signal

import (
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	signalingpb "github.com/peerrpc/go/gen/proto/peerrpc/signaling"

	"github.com/gorilla/websocket"
	"google.golang.org/protobuf/proto"
)

// WSOption configures a WS backend.
type WSOption func(*WS)

// WithToken supplies a bearer token forwarded to the signal-server.
// The token is sent as a ?token= query parameter on the WebSocket
// handshake URL (matching the server's WebSocketHandler) and as an
// Authorization header fallback.
func WithToken(token string) WSOption {
	return func(w *WS) { w.token = token }
}

// WithDialer overrides the default websocket.Dialer (e.g. to set a
// custom TLS config or timeouts).
func WithDialer(d *websocket.Dialer) WSOption {
	return func(w *WS) { w.dialer = d }
}

// WithHandshakeTimeout sets the WebSocket handshake timeout.
func WithHandshakeTimeout(d time.Duration) WSOption {
	return func(w *WS) { w.handshakeTimeout = d }
}

// WS is a Backend that connects to a remote signal-server over a
// WebSocket. The constructor takes a base URL (any of http(s):// or
// ws(s):// forms; a bare host is also accepted) and optional WSOption
// values.
//
// Usage:
//
//	backend := signal.NewWS("ws://localhost:8443")
//	sig, err := backend.Exchange(ctx, "service-id", "peer-id")
//
// WS is safe for concurrent use by any number of peers and services.
type WS struct {
	baseURL          string
	token            string
	dialer           *websocket.Dialer
	handshakeTimeout time.Duration
}

// NewWS constructs a WS backend pointing at the given signal-server.
// The URL is normalized: http(s):// is rewritten to ws(s):// and a
// "/ws" path is appended when absent.
func NewWS(baseURL string, opts ...WSOption) *WS {
	w := &WS{baseURL: baseURL}
	for _, opt := range opts {
		opt(w)
	}
	return w
}

// wsURL normalizes the configured base URL into a ws:// or wss://
// URL ending with the "/ws" path, appending the token query param
// when a token is set.
func (w *WS) wsURL() (string, error) {
	raw := strings.TrimSpace(w.baseURL)
	if raw == "" {
		return "", errors.New("signal: empty ws url")
	}

	// Rewrite the scheme: http(s) → ws(s). A bare host (no scheme)
	// defaults to ws://.
	var u *url.URL
	var err error
	switch {
	case strings.HasPrefix(raw, "http://"):
		u, err = url.Parse("ws://" + raw[len("http://"):])
	case strings.HasPrefix(raw, "https://"):
		u, err = url.Parse("wss://" + raw[len("https://"):])
	case strings.HasPrefix(raw, "ws://"), strings.HasPrefix(raw, "wss://"):
		u, err = url.Parse(raw)
	default:
		// Bare host[:port][/path]. Treat as cleartext ws://.
		u, err = url.Parse("ws://" + raw)
	}
	if err != nil {
		return "", fmt.Errorf("signal: parse ws url %q: %w", raw, err)
	}

	// Append the "/ws" path if none (or only "/") is present.
	if u.Path == "" || u.Path == "/" {
		u.Path = "/ws"
	}

	// Carry the token as a query param (the server's WS auth reads it
	// from here first).
	if w.token != "" {
		q := u.Query()
		q.Set("token", w.token)
		u.RawQuery = q.Encode()
	}
	return u.String(), nil
}

// Exchange implements Backend by opening a WebSocket to the
// signal-server. The first frame sent is an AnnounceRequest carrying
// the peer id and role. Subsequent outbound messages are forwarded
// verbatim; inbound messages from the server are delivered via the
// returned Session's Receive channel.
//
// The caller MUST close the Session when done (or cancel ctx) to
// release the connection.
func (w *WS) Exchange(ctx context.Context, service, peerID string) (*Session, error) {
	if service == "" {
		return nil, errors.New("signal: empty service")
	}
	if peerID == "" {
		return nil, errors.New("signal: empty peer id")
	}

	wsURL, err := w.wsURL()
	if err != nil {
		return nil, err
	}

	dialer := w.dialer
	if dialer == nil {
		dialer = &websocket.Dialer{
			HandshakeTimeout: w.handshakeTimeout,
		}
		if w.handshakeTimeout == 0 {
			dialer.HandshakeTimeout = 10 * time.Second
		}
	}

	header := http.Header{}
	if w.token != "" {
		header.Set("Authorization", "Bearer "+w.token)
	}

	conn, _, err := dialer.DialContext(ctx, wsURL, header)
	if err != nil {
		return nil, fmt.Errorf("signal: ws dial %q: %w", wsURL, err)
	}

	// First message MUST be AnnounceRequest.
	if err := writeFrame(conn, &signalingpb.SignalMessage{
		Service: service,
		Body: &signalingpb.SignalMessage_Announce{
			Announce: &signalingpb.AnnounceRequest{
				PeerId: peerID,
				Role:   signalingpb.AnnounceRequest_ROLE_CLIENT,
			},
		},
	}); err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("signal: send announce: %w", err)
	}

	in := make(chan *SignalMessage, 32)
	out := make(chan *SignalMessage, 32)

	ctx, cancel := context.WithCancel(ctx)

	// A dedicated write mutex serializes concurrent frame writes (the
	// gorilla websocket.Conn.WriteMessage is not safe for concurrent
	// callers).
	var writeMu sync.Mutex

	// Pump: outbound channel -> ws.WriteMessage.
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			select {
			case msg, ok := <-out:
				if !ok {
					// Half-close: signal end-of-stream to the server.
					writeMu.Lock()
					_ = conn.WriteMessage(websocket.CloseMessage,
						websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
					writeMu.Unlock()
					return
				}
				pb, err := toProtoMessage(msg)
				if err != nil {
					continue
				}
				pb.Service = service
				writeMu.Lock()
				werr := writeFrame(conn, pb)
				writeMu.Unlock()
				if werr != nil {
					return
				}
			case <-ctx.Done():
				return
			}
		}
	}()

	// Pump: ws.ReadMessage -> inbound channel.
	wg.Add(1)
	go func() {
		defer wg.Done()
		defer cancel()
		for {
			pb, err := readFrame(conn)
			if err != nil {
				close(in)
				return
			}
			sm := fromProtoMessage(pb)
			if sm == nil {
				continue
			}
			select {
			case in <- sm:
			default:
				// Inbox full; drop (best-effort signaling).
			}
		}
	}()

	s := &Session{
		service: service,
		peerID:  peerID,
		out:     out,
		in:      in,
		done:    make(chan struct{}),
		cleanup: func() {
			cancel()
			_ = conn.Close()
		},
	}
	return s, nil
}

// readFrame reads one 4-byte-length-prefixed SignalMessage from the
// WebSocket (BinaryMessage only).
func readFrame(conn *websocket.Conn) (*signalingpb.SignalMessage, error) {
	_, data, err := conn.ReadMessage()
	if err != nil {
		return nil, err
	}
	if len(data) < 4 {
		return nil, fmt.Errorf("ws frame too short: %d bytes", len(data))
	}
	length := binary.BigEndian.Uint32(data[:4])
	if uint32(len(data)-4) < length {
		return nil, fmt.Errorf("ws frame truncated: want %d payload bytes, got %d", length, len(data)-4)
	}
	var msg signalingpb.SignalMessage
	if err := proto.Unmarshal(data[4:4+length], &msg); err != nil {
		return nil, fmt.Errorf("ws unmarshal: %w", err)
	}
	return &msg, nil
}

// writeFrame writes one 4-byte-length-prefixed SignalMessage to the
// WebSocket as a BinaryMessage.
func writeFrame(conn *websocket.Conn, msg *signalingpb.SignalMessage) error {
	payload, err := proto.Marshal(msg)
	if err != nil {
		return fmt.Errorf("ws marshal: %w", err)
	}
	frame := make([]byte, 4+len(payload))
	binary.BigEndian.PutUint32(frame[:4], uint32(len(payload)))
	copy(frame[4:], payload)
	return conn.WriteMessage(websocket.BinaryMessage, frame)
}

// toProtoMessage converts a signal.SignalMessage to the protobuf
// wire type.
func toProtoMessage(msg *SignalMessage) (*signalingpb.SignalMessage, error) {
	pb := &signalingpb.SignalMessage{}
	switch {
	case msg.Body.Announce != nil:
		a := msg.Body.Announce
		req := &signalingpb.AnnounceRequest{
			PeerId: a.PeerID,
			Role:   signalingpb.AnnounceRequest_Role(a.Role),
		}
		if a.PeerPubkey != nil {
			req.PeerPubkey = a.PeerPubkey
		}
		pb.Body = &signalingpb.SignalMessage_Announce{Announce: req}
	case msg.Body.Offer != nil:
		pb.Body = &signalingpb.SignalMessage_Offer{
			Offer: &signalingpb.SdpOffer{Sdp: msg.Body.Offer.Sdp},
		}
	case msg.Body.Answer != nil:
		pb.Body = &signalingpb.SignalMessage_Answer{
			Answer: &signalingpb.SdpAnswer{Sdp: msg.Body.Answer.Sdp},
		}
	case msg.Body.Candidate != nil:
		pb.Body = &signalingpb.SignalMessage_Candidate{
			Candidate: &signalingpb.IceCandidate{
				Candidate:     msg.Body.Candidate.Candidate,
				SdpMid:        msg.Body.Candidate.SdpMid,
				SdpMlineIndex: msg.Body.Candidate.SdpMLineIndex,
			},
		}
	case msg.Body.Leave != nil:
		pb.Body = &signalingpb.SignalMessage_Leave{
			Leave: &signalingpb.LeaveRequest{Reason: msg.Body.Leave.Reason},
		}
	case msg.Body.Ping != nil:
		pb.Body = &signalingpb.SignalMessage_Ping{
			Ping: &signalingpb.Ping{TimestampMs: msg.Body.Ping.TimestampMs},
		}
	default:
		return nil, errors.New("signal: unknown body type")
	}
	return pb, nil
}

// fromProtoMessage converts a protobuf SignalMessage to the
// signal.SignalMessage type. Returns nil for unrecognised bodies.
func fromProtoMessage(pb *signalingpb.SignalMessage) *SignalMessage {
	sm := &SignalMessage{Service: pb.GetService()}
	switch body := pb.GetBody().(type) {
	case *signalingpb.SignalMessage_Announce:
		sm.Body.Announce = &AnnounceRequest{
			PeerID: body.Announce.GetPeerId(),
			Role:   Role(body.Announce.GetRole()),
		}
		if body.Announce.GetPeerPubkey() != nil {
			sm.Body.Announce.PeerPubkey = body.Announce.GetPeerPubkey()
		}
	case *signalingpb.SignalMessage_Offer:
		sm.Body.Offer = &SdpOffer{Sdp: body.Offer.GetSdp()}
	case *signalingpb.SignalMessage_Answer:
		sm.Body.Answer = &SdpAnswer{Sdp: body.Answer.GetSdp()}
	case *signalingpb.SignalMessage_Candidate:
		sm.Body.Candidate = &IceCandidate{
			Candidate:     body.Candidate.GetCandidate(),
			SdpMid:        body.Candidate.GetSdpMid(),
			SdpMLineIndex: body.Candidate.GetSdpMlineIndex(),
		}
	case *signalingpb.SignalMessage_Leave:
		sm.Body.Leave = &LeaveRequest{Reason: body.Leave.GetReason()}
	case *signalingpb.SignalMessage_Ping:
		sm.Body.Ping = &Ping{TimestampMs: body.Ping.GetTimestampMs()}
	default:
		return nil
	}
	return sm
}

// Ensure WS satisfies Backend at compile time.
var _ Backend = (*WS)(nil)

File: peerrpc-rs/signal/src/ws.rs
//! WebSocket backend: a signaling client that speaks the signaling
//! wire format (service / AnnounceRequest) to a remote signal-server
//! over a raw WebSocket.
//!
//! Mirrors Go's `signal.WS` and TS's `WebSocketSignal`. The server's
//! `/ws` endpoint expects each frame to be a 4-byte big-endian length
//! prefix + a marshaled `peerrpc.signaling.SignalMessage` protobuf, with
//! the first frame an `AnnounceRequest` carrying (service, peer_id, role).

use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use prost::Message as _;
use tokio::sync::{mpsc, oneshot, Notify};
use tokio_tungstenite::tungstenite::Message;

use peerrpc_protocol::gen::signaling::{
    announce_request, signal_message, AnnounceRequest, SignalMessage as WireMsg,
};

use crate::{IceCandidate, SdpAnswer, SdpOffer, Session, SignalBody, SignalError, SignalMessage};

/// `WS` is a signaling backend that connects to a remote signal-server
/// over a WebSocket. Construct with a base URL (any of http(s):// or
/// ws(s):// forms; a bare host is also accepted) and call `exchange`
/// per peer.
///
/// The URL is normalized: http(s):// is rewritten to ws(s):// and a
/// "/ws" path is appended when absent, matching the server's endpoint.
pub struct WS {
    base_url: String,
    token: Option<String>,
}

impl WS {
    /// Construct a WS backend pointing at the given signal-server. The
    /// URL is normalized on each `exchange`.
    pub fn new(base_url: impl Into<String>) -> Self {
        Self {
            base_url: base_url.into(),
            token: None,
        }
    }

    /// Supply a bearer token forwarded to the signal-server (as a
    /// `?token=` query param on the WebSocket URL).
    pub fn with_token(mut self, token: impl Into<String>) -> Self {
        self.token = Some(token.into());
        self
    }

    /// Open a WebSocket to the signal-server, send the AnnounceRequest
    /// as the first frame, and return a Session that ferries
    /// SignalMessages both ways. Each call dials its own WebSocket.
    pub async fn exchange(&self, service: &str, peer_id: &str) -> Result<Session, SignalError> {
        if service.is_empty() {
            return Err(SignalError::Other("empty service".into()));
        }
        if peer_id.is_empty() {
            return Err(SignalError::Other("empty peer id".into()));
        }

        let url = ws_url(&self.base_url, self.token.as_deref())?;

        // Dial the WebSocket; surface the result via a oneshot so a
        // connect failure is not silently swallowed.
        let (connect_tx, connect_rx) = oneshot::channel::<Result<(), String>>();
        let (mut ws_stream, _resp) = match tokio_tungstenite::connect_async(url.clone()).await {
            Ok(pair) => {
                let _ = connect_tx.send(Ok(()));
                pair
            }
            Err(e) => {
                return Err(SignalError::Other(format!("ws dial {url:?}: {e}")));
            }
        };

        // First frame MUST be the AnnounceRequest.
        let announce = WireMsg {
            service: service.to_string(),
            body: Some(signal_message::Body::Announce(AnnounceRequest {
                peer_id: peer_id.to_string(),
                peer_pubkey: None,
                role: announce_request::Role::Server as i32,
            })),
        };
        if let Err(e) = ws_stream.send(Message::Binary(write_frame(&announce))).await {
            return Err(SignalError::Other(format!("ws send announce: {e}")));
        }

        let svc = service.to_string();
        let peer = peer_id.to_string();
        let done = Arc::new(Notify::new());
        let done_for_pump = done.clone();

        // Public channels bridged to the wire.
        let (pub_outbound_tx, mut pub_outbound_rx) = mpsc::unbounded_channel::<SignalMessage>();
        let (pub_inbound_tx, pub_inbound_rx) = mpsc::unbounded_channel::<SignalMessage>();

        // One pump task owns the WebSocket and drives both directions.
        // Outbound: public SignalMessage -> wire frame -> ws.send.
        // Inbound:  ws.next -> wire frame -> public SignalMessage.
        tokio::spawn(async move {
            // connect_tx is already consumed above; drop the stale
            // receiver pattern by moving nothing extra in.
            loop {
                tokio::select! {
                    biased;
                    _ = done_for_pump.notified() => break,
                    msg = pub_outbound_rx.recv() => {
                        let Some(msg) = msg else { break };
                        let wire = translate_out(&msg);
                        let frame = write_frame(&wire);
                        if ws_stream.send(Message::Binary(frame)).await.is_err() {
                            break;
                        }
                    }
                    frame = ws_stream.next() => {
                        match frame {
                            Some(Ok(Message::Binary(data))) => {
                                if let Some(wire) = read_frame(&data) {
                                    let translated = translate_in(&wire);
                                    if pub_inbound_tx.send(translated).is_err() {
                                        break;
                                    }
                                }
                            }
                            Some(Ok(_)) => {
                                // Ignore text/ping/close-control frames.
                            }
                            Some(Err(e)) => {
                                tracing::warn!("signal: ws read error: {e}");
                                break;
                            }
                            None => break, // server closed
                        }
                    }
                }
            }
            let _ = ws_stream.close(None).await;
        });

        // Surface any deferred connect error (currently always Ok since
        // we awaited connect_async above, but kept for symmetry with
        // Remote and future lazy-connect variants).
        match connect_rx.await {
            Ok(Ok(())) => {}
            Ok(Err(msg)) => return Err(SignalError::Other(msg)),
            Err(_) => return Err(SignalError::Other("signal: connect task dropped".into())),
        }

        Ok(Session {
            service: svc,
            peer_id: peer,
            outbound: pub_outbound_tx,
            inbound: pub_inbound_rx,
            done,
            cleanup: None,
        })
    }
}

// ── URL normalization ──────────────────────────────────────────

/// Normalize the configured base URL into a ws:// or wss:// URL ending
/// with the "/ws" path, appending the token query param when set.
fn ws_url(base: &str, token: Option<&str>) -> Result<String, SignalError> {
    let raw = base.trim();
    if raw.is_empty() {
        return Err(SignalError::Other("empty ws url".into()));
    }
    // Rewrite scheme: http(s) -> ws(s). Bare host defaults to ws://.
    let rewritten = if let Some(rest) = raw.strip_prefix("http://") {
        format!("ws://{rest}")
    } else if let Some(rest) = raw.strip_prefix("https://") {
        format!("wss://{rest}")
    } else if raw.starts_with("ws://") || raw.starts_with("wss://") {
        raw.to_string()
    } else {
        format!("ws://{raw}")
    };
    let mut url: http::Uri = rewritten
        .parse()
        .map_err(|e| SignalError::Other(format!("parse ws url {raw:?}: {e}")))?;

    // Append "/ws" path when none (or only "/") is present.
    let path = url.path();
    if path.is_empty() || path == "/" {
        // Rebuild with /ws path (http::Uri is partially mutable via path().
        let mut parts = url.into_parts();
        parts.path_and_query = Some(format!(
            "/ws{}",
            token
                .map(|t| format!("?token={}", urlenc(t)))
                .unwrap_or_default()
        )
        .parse()
        .map_err(|e| SignalError::Other(format!("rebuild ws url: {e}")))?);
        url = http::Uri::from_parts(parts)
            .map_err(|e| SignalError::Other(format!("rebuild ws url: {e}")))?;
    } else if let Some(t) = token {
        // Path already present; append token as query.
        let existing = url.path_and_query().map(|pq| pq.as_str()).unwrap_or(path);
        let sep = if existing.contains('?') { '&' } else { '?' };
        let with_token = format!("{existing}{sep}token={}", urlenc(t));
        let mut parts = url.into_parts();
        parts.path_and_query = Some(
            with_token
                .parse()
                .map_err(|e| SignalError::Other(format!("rebuild ws url: {e}")))?,
        );
        url = http::Uri::from_parts(parts)
            .map_err(|e| SignalError::Other(format!("rebuild ws url: {e}")))?;
    }
    Ok(url.to_string())
}

fn urlenc(s: &str) -> String {
    // Minimal percent-encoding for token query values.
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'-' | b'.' | b'_' | b'~' | b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{:02X}", b)),
        }
    }
    out
}

// ── Wire framing (4-byte BE length prefix + protobuf) ──────────

/// Encode a SignalMessage into a 4-byte BE length prefix + protobuf.
fn write_frame(msg: &WireMsg) -> Vec<u8> {
    let payload = msg.encode_to_vec();
    let mut out = Vec::with_capacity(4 + payload.len());
    out.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    out.extend_from_slice(&payload);
    out
}

/// Decode one 4-byte-length-prefixed SignalMessage from a WebSocket
/// Binary frame. Returns None if the frame is malformed.
fn read_frame(data: &[u8]) -> Option<WireMsg> {
    if data.len() < 4 {
        return None;
    }
    let len = u32::from_be_bytes([data[0], data[1], data[2], data[3]]) as usize;
    if data.len() - 4 < len {
        return None;
    }
    WireMsg::decode(&data[4..4 + len]).ok()
}

// ── Wire <-> public type translation ───────────────────────────

fn translate_out(msg: &SignalMessage) -> WireMsg {
    let body = match &msg.body {
        SignalBody {
            offer: Some(o), ..
        } => Some(signal_message::Body::Offer(peerrpc_protocol::gen::signaling::SdpOffer {
            sdp: o.sdp.clone(),
        })),
        SignalBody {
            answer: Some(a), ..
        } => Some(signal_message::Body::Answer(peerrpc_protocol::gen::signaling::SdpAnswer {
            sdp: a.sdp.clone(),
        })),
        SignalBody {
            candidate: Some(c), ..
        } => Some(signal_message::Body::Candidate(peerrpc_protocol::gen::signaling::IceCandidate {
            candidate: c.candidate.clone(),
            sdp_mid: c.sdp_mid.clone(),
            sdp_mline_index: c.sdp_m_line_index,
        })),
        _ => None,
    };
    WireMsg {
        service: msg.service.clone(),
        body,
    }
}

fn translate_in(wire: &WireMsg) -> SignalMessage {
    let body = match &wire.body {
        Some(signal_message::Body::Announce(_)) => SignalBody::default(),
        Some(signal_message::Body::Offer(o)) => SignalBody {
            offer: Some(SdpOffer { sdp: o.sdp.clone() }),
            ..Default::default()
        },
        Some(signal_message::Body::Answer(a)) => SignalBody {
            answer: Some(SdpAnswer { sdp: a.sdp.clone() }),
            ..Default::default()
        },
        Some(signal_message::Body::Candidate(c)) => SignalBody {
            candidate: Some(IceCandidate {
                candidate: c.candidate.clone(),
                sdp_mid: c.sdp_mid.clone(),
                sdp_m_line_index: c.sdp_mline_index,
            }),
            ..Default::default()
        },
        _ => SignalBody::default(),
    };
    SignalMessage {
        service: wire.service.clone(),
        body,
    }
}
